[CSSimplify] Skip transitive conformance check if argument is `CGFloat` or `Double`

xedin opened the pull request:
There is an implicit conversion between `CGFloat` and `Double` types 
so if `CGFloat`/`Double` argument doesn't satisfy the requirement it's
possible that the generic parameter it's passed to be inferred as a the
other type from context and the requirement to be satisfied through this 
implicit conversion.

Resolves: rdar://153461854

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSSimplify] Skip transitive conformance check if argument is `CGFloat`

There is an implicit conversion between `CGFloat` and `Double` types
so if `CGFloat` argument doesn't satisfy the requirement it's possible
that the generic parameter it's passed to be inferred as a `Double`
from context and the requirement to be satisfied through this implicit
conversion.

Resolves: rdar://153461854

xedin commented:
@swift-ci please test macOS platform

hamishknight commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -9247,6 +9247,14 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyTransitivelyConformsTo(
   if (resolvedTy->isTypeVariableOrMember())
     return formUnsolved();
 
+  // There is an implicit conversion between `CGFloat` and `Double` types
+  // so if `CGFloat` argument doesn't satisfy the requirement it's possible
+  // that the generic parameter it's passed to be inferred as a `Double`
+  // from context and the requirement to be satisfied through this implicit
+  // conversion.
+  if (resolvedTy->isCGFloat())
+    return SolutionKind::Solved;
```
Isn't this also an issue going from Double to CGFloat? e.g:

```swift
import CoreGraphics

protocol P {}
extension CGFloat: P {}

struct S<T> {}

func foo<T: P>(_ x: T) -> S<T> { .init() }

func bar(_ x: Double, _ y: CGFloat) {
  _ = {
    let x: S<CGFloat> = foo(x)
  }
}
```

xedin replied:
Yeah, I didn't include it because I'm afraid to cause perf regressions because `Double` is more widely used.

xedin replied:
Let's try that as a separate commit and we can revert if that doesn't pan out.

xedin committed:
[CSSimplify] Skip transitive conformance check if argument is `Double`

This is the same problem as `CGFloat` because the conversion is
bi-directional.

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please clean test macOS platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test source compatibility debug

xedin commented:
This has performance implications even if we reduce it only to CGFloat, I'll fix it in a different way for now.
